lib/rubocop/cop/lint/reduce_accumulator.rb

```diff
@@ -0,0 +1,148 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Lint
+      # Looks for `reduce` or `inject` blocks where the value returned (implicitly or
+      # explicitly) does not include the accumulator. A block is considered valid as
+      # long as at least one return value includes the accumulator.
+      #
+      # If the accumulator is not included in the return value, then the entire
+      # block will just return a transformation of the last element value, and
+      # could be rewritten as such without a loop.
+      #
+      # Also catches instances where an index of the accumulator is returned, as
+      # this changes the type of object being retained.
+      #
+      # NOTE: For the purpose of reducing false positives, this cop only flags
+      # `reduce` blocks where an expression
+      #
+      # @example
+      #
+      #   # bad
+      #   (1..4).reduce(0) do |acc, i|
+      #     i * 2
+      #   end
+      #
+      #   # bad, may raise a NoMethodError after the first iteration
+      #   %w(a b c).reduce({}) do |acc, letter|
+      #     acc[letter] = true
+      #   end
+      #
+      #   # good
+      #   (1..4).reduce(0) do |acc, i|
+      #     acc + i * 2
+      #   end
+      #
+      #   %w(a b c).reduce({}) do |acc, letter|
+      #     acc[letter] = true
+      #     acc
+      #   end
+      class ReduceAccumulator < Base
+        MSG = 'Ensure the accumulator will be modified by `%<method>s`.'
```

Comment:
Might be a good idea to mention the accumulator name explicitly in the messages.